File: app/src/main/java/org/ngbp/jsonrpc4jtestharness/rsp/receiverQueryApi/api/ReceiverQueryApiImpl.java
package org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.api;

import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.Alerting;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.AudioAccessibilityPref;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.BaseURI;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.CC;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.CaptionDisplay;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.Languages;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.RatingLevel;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.Service;
import org.ngbp.jsonrpc4jtestharness.rsp.receiverQueryApi.model.ServiceGuideUrls;

import java.util.List;

public class ReceiverQueryApiImpl implements ReceiverQueryApi {

    @Override
    public RatingLevel queryContentAdvisoryRating() {
        return null;
    }

    @Override
    public CC queryClosedCaptionsStatus() {
        return null;
    }

    @Override
    public Service queryServiceID() {
        return null;
    }

    @Override
    public Languages queryLanguagePreferences() {
        return null;
    }

    @Override
    public CaptionDisplay queryCaptionDisplayPreferences() {
        return null;
    }

    @Override
    public AudioAccessibilityPref queryAudioAccessibilityPreferences() {
        return null;
    }

    @Override
    public BaseURI queryReceiverWebServerURI() {
        return null;
    }

    @Override
    public List<Alerting> queryAlertingSignaling() {
        return null;
    }

    @Override
    public List<ServiceGuideUrls> queryServiceGuideURLs() {
        return null;
    }
}

File: nextgen-middleware/src/main/java/com/nextgenbroadcast/mobile/middleware/rpc/receiverQueryApi/ReceiverQueryApiImpl.kt
package com.nextgenbroadcast.mobile.middleware.rpc.receiverQueryApi

import com.nextgenbroadcast.mobile.middleware.atsc3.Atsc3LLSTable
import com.nextgenbroadcast.mobile.middleware.rpc.RpcException
import com.nextgenbroadcast.mobile.middleware.rpc.receiverQueryApi.model.*
import com.nextgenbroadcast.mobile.middleware.server.IApplicationSession

class ReceiverQueryApiImpl(
    private val session: IApplicationSession
) : IReceiverQueryApi {

    override fun queryContentAdvisoryRating(): RatingLevelRpcResponse {
        return RatingLevelRpcResponse()
    }

    override fun queryClosedCaptionsStatus(): CaptionsRpcResponse {
        return CaptionsRpcResponse()
    }

    override fun queryServiceID(): ServiceRpcResponse {
        return ServiceRpcResponse().apply {
            this.service = session.getParam(IApplicationSession.Params.ServiceId)
        }
    }

    override fun queryLanguagePreferences(): LanguagesRpcResponse {
        val language = session.getParam(IApplicationSession.Params.Language)
        return LanguagesRpcResponse().apply {
            preferredAudioLang = language
            preferredCaptionSubtitleLang = language
            preferredUiLang = language
        }
    }

    override fun queryCaptionDisplayPreferences(): CaptionDisplayRpcResponse {
        return CaptionDisplayRpcResponse()
    }

    override fun queryAudioAccessibilityPreferences(): AudioAccessibilityPrefRpcResponse {
        return AudioAccessibilityPrefRpcResponse()
    }

    override fun queryMPDUrl(): MPDUrlRpcResponse {
        return MPDUrlRpcResponse(session.getParam(IApplicationSession.Params.MediaUrl))
    }

    override fun queryReceiverWebServerURI(): BaseURIRpcResponse {
        return session.getParam(IApplicationSession.Params.AppBaseUrl)?.let {
            BaseURIRpcResponse(it)
        } ?: throw RpcException()
    }

    override fun queryAlertingSignaling(alertingTypes: List<String>): AlertingSignalingRpcResponse {
        val aeatList = if (alertingTypes.contains(AlertingSignalingRpcResponse.Alert.AEAT)) {
            session.getAEATChangingList()
        } else emptyList()
        return AlertingSignalingRpcResponse(
            listOf(AlertingSignalingRpcResponse.Alert(AlertingSignalingRpcResponse.Alert.AEAT,
                aeatList.joinToString(separator = "", prefix = "<AEAT>", postfix = "</AEAT>") { it }))
        )
    }

    override fun queryServiceGuideURLs(service: String?): ServiceGuideUrlsRpcResponse {
        return ServiceGuideUrlsRpcResponse(

            session.getServiceGuideUrls(service).map { sgUrl ->
                ServiceGuideUrlsRpcResponse.Url(
                    sgUrl.sgType.toString(),
                    sgUrl.sgPath,
                    sgUrl.service,
                    sgUrl.content
                )
            }
        )
    }

    override fun querySignaling(names: List<String>): SignalingRpcResponse {
        return SignalingRpcResponse(
            session.getSignalingInfo(names).map { data ->
                SignalingRpcResponse.SignalingInfo(
                    name = data.name,
                    group = (data as? Atsc3LLSTable)?.groupId?.toString(),
                    version = data.version.toString(),
                    table = data.xml
                )
            }
        )
    }

}